sdk/core/Azure.Core.TestFramework/src/Instrumentation/DiagnosticScopeValidatingInterceptor.cs

```diff
@@ -260,6 +261,58 @@ internal static async ValueTask<T> ValidateDiagnosticScope<T>(Func<ValueTask<(T
+                if (!Regex.IsMatch(tag.Key, @"^[a-z\._]+$"))
+                {
+                    throw new InvalidOperationException("Attribute name can only have lowercase letters, dot (`.`), and underscore (`_`). " +
+                        "Use dot to separate namespaces and underscore to separate words (e.g. http.request.status_code). " + $"Attribute name: {tag.Key}");
+                }
+
+                int dot = tag.Key.IndexOf('.');
+                if (dot == -1)
+                {
+                    throw new InvalidOperationException("Attribute names must be namespaced. Use OpenTelemetry attributes whenever possible - https://github.com/open-telemetry/semantic-conventions/blob/main/docs/README.md. " +
+                        "Custom Azure-specific attributes must start with `az.` and should have library-specific namespaces (e.g. `az.digital_twin.twin_id`). " + $"Attribute name: {tag.Key}");
+                }
+
+                string ns = tag.Key.Substring(0, dot);
+                if ("az" != ns && "messaging" != ns && "http" != ns && "error" != ns && "url" != ns)
+                {
+                    throw new InvalidOperationException("Unknown attribute namespace. Use OpenTelemetry attributes whenever possible - https://github.com/open-telemetry/semantic-conventions/blob/main/docs/README.md." +
+                        "Custom Azure-specific attributes must start with `az.` and should have library-specific namespaces (e.g. `az.digital_twin.twin_id`). " + $"Attribute name: {tag.Key}");
+                }
+
+                string tagValueStr = tag.Value?.ToString();
+                if (string.IsNullOrEmpty(tagValueStr) || (tag.Key.StartsWith("az.") && tagValueStr.Length > 256))
+                {
+                    throw new InvalidOperationException("Attribute values must not be null, empty, or too long. "
+                        + $"Attribute name: `{tag.Key}`, value: `{tag.Value}`, activity name: `{activity.OperationName}`");
+                }
+            }
+
+            if (strict)
+            {
+                if (!activity.Tags.Any(tag => tag.Key == "az.namespace"))
+                {
+                    throw new InvalidOperationException($"All diagnostic scopes should have 'az.namespace' attribute, make sure the assembly containing **ClientOptions type is marked with the AzureResourceProviderNamespace attribute specifying the appropriate provider. This attribute should be included in AssemblyInfo, and can be included by pulling in AzureResourceProviderNamespaceAttribute.cs using the AzureCoreSharedSources alias.");
+                }
+            }
+
+            if (activity.Status == ActivityStatusCode.Error && activity.Source?.HasListeners() == true && !activity.TagObjects.Any(kvp => kvp.Key == "error.type"))
+            {
+                throw new InvalidOperationException("All failed activities must have `error.type` attribute set to low-cardinality error code or a full name of exception type");
```

Comment:
Will consumers of this error know what it means to set a low-cardinality error code (I'm not sure I do 😄 ). Perhaps we could be more prescriptive in what that looks like?